kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/DeltaTableReadsSuite.scala

```diff
@@ -1063,4 +1066,161 @@ class DeltaTableReadsSuite extends AnyFunSuite with TestUtils {
       }
     }
   }
+
+  def testWithCompactions(
+      versionsToWrite: Seq[Int], // highest version MUST be last!
+      versionToRead: Option[Long],
+      doRemoves: Boolean,
+      compactions: Seq[(Int, Int)],
+      expectedDeltasToBeRead: Set[Int],
+      expectedCompactionsToBeRead: Set[(Int, Int)]) {
+    withTempDir { tmpDir =>
+      val tablePath = tmpDir.getCanonicalPath
+      val hadoopFileIO = new HadoopFileIO(new Configuration() {
+        {
+          // Set the batch sizes to small so that we get to test the multiple batch scenarios.
+          set("delta.kernel.default.parquet.reader.batch-size", "2");
+          set("delta.kernel.default.json.reader.batch-size", "2");
+        }
+      })
+      val engine = new MetricsEngine(hadoopFileIO)
+      var expectedRows: Set[Long] = Set()
+      versionsToWrite.foreach { i =>
+        // if we're removing, then on odd commits, remove the lower 10 of the previous 20 rows added
+        if (doRemoves && i % 2 == 1) {
+          val prev = i - 1;
+          val low = prev * 10
+          val high = prev * 10 + 10
+          val deleteQuery = "DELETE FROM delta.`%s` WHERE id >= %d AND id < %d".format(
+            tablePath,
+            low,
+            high)
+          spark.sql(deleteQuery)
+          if (versionToRead.isEmpty || versionToRead.get >= i) {
+            expectedRows --= (low until high).map(i => i.toLong)
+          }
+        } else {
+          val low = i * 10
+          // if we're removing, add 20 rows as the first 10 will be removed by the next version,
+          // otherwise add 10 rows
+          val high = if (doRemoves) low + 20 else low + 10
+          spark.range(low, high).write
+            .format("delta")
+            .mode("append")
+            .save(tablePath)
+
+          if (versionToRead.isEmpty || versionToRead.get >= i) {
+            expectedRows ++= (low until high).map(i => i.toLong)
+          }
+        }
+      }
+
+      val dataPath = new Path(s"file:${tablePath}")
+      val logPath = new Path(s"file:${tablePath}", "_delta_log")
+      // create the compaction file(s)
+      compactions.foreach { compaction =>
+        val hook = new LogCompactionHook(
+          dataPath,
+          logPath,
+          compaction._1,
+          compaction._2,
+          0)
+        hook.threadSafeInvoke(engine)
+      }
+      engine.resetMetrics()
+
+      checkTable(
```

Comment:
We are using the log compactions for three different operations right? (1) snapshot loading (P&M only) (2) log replay (Scans) (3) loading the metadata domain map

Should we test that we use it for each of these? It seems like maybe before these recent commits we actually had missed (2)?

I think this doesn't necessarily need to be too complex of a test, we can test all of these things using 1 single test table built with just a few versions (and reset the metrics in between each check).

Reply:
Alternatively maybe we can accomplish this instead by not converting the `actualCompactionsRead` to set and instead checking we actually read a specific compaction file twice for example when snapshot laoding + scanning

Reply:
Yep. I've moved this to a new test file now, and added a test with DM and a property change.

I didn't include removes in that test because I don't think kernel can do that just yet (easily), but that's well covered by the other tests.